### PR TITLE
QualityAnalyzer plugin

### PR DESCRIPTION
**Pull request by cg122**
## Description

QualityAnalyzer is FASTEN plugin to store/update code quality metadata from Kafka topic. There is a single Kafka message per single callable produced by [Rapid Plugin](https://github.com/fasten-project/quality-analyzer/tree/master/rapidplugin) using [Lizard](https://github.com/terryyin/lizard). Once a metadata message is consumed, the metadata is stored to respective database depending on the forge (mvn, c, python).

## Motivation and context

One of the use-case scenarios within FASTEN project is application level risk profile with external dependencies. The application-level risk considering a list of claimed transitive dependencies and this evaluation may cover code quality. Hence, a need to perform code quality analysis and to store the results.

## Testing the plugin

```
java -cp /path/to/projects/fasten/docker/server/server-0.0.1-SNAPSHOT-with-dependencies.jar eu.fasten.server.FastenServer -p /home/fasten/projects/fasten/docker/plugins/ -k localhost:9092 -pl "QualityAnalyzer" -kt "QualityAnalyzer=fasten.RapidPlugin.out" -d "mvn=jdbc:postgresql://postgres@localhost/dbname"
```
## Task list  

- [x] Create a plugin to consume topic produced by RapidPlugin
- [x] Create DB connection to the metadata DB
- [x]  Write query to do the DB updating for file/callable level metrics
- [x] Create test cases
- [x] Update documentation

### CODE REVIEW

File: analyzer/quality-analyzer/pom.xml
Comment (MihhailSokolov) [inline analyzer/quality-analyzer/pom.xml:56]: Please take a look at other plugins' pom files. We updated the dependency handling so now you don't need to comment/uncomment some dependencies to run the Main class

File: analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/QualityAnalyzerPlugin.java
Comment (MihhailSokolov) [inline analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/QualityAnalyzerPlugin.java:43]: `transactionRestartLimit` is now a part of Constants class in `core/data/Constants`, so please reference it from there

File: analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/QualityAnalyzerPlugin.java
Comment (MihhailSokolov) [inline analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/QualityAnalyzerPlugin.java:94]: `":"` is also one of the Constants. Please reference it as `Constants.mvnCoordinateSeparator` everywhere instead of hard-coding `":"`

File: analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/QualityAnalyzerPlugin.java
Comment (MihhailSokolov) [inline analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/QualityAnalyzerPlugin.java:120]: `"mvn"` is also a Constant :)
`Constants.mvnForge`
In general if you have any hard-coded constants, please put them in the Constants class

File: analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/QualityAnalyzerPlugin.java
Comment (MihhailSokolov) [inline analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/QualityAnalyzerPlugin.java:135]: Please add a proper output path like `groupId + "_" + artifactId + "_" + version + ".json"` You can take a look at what other plugins have as the output path.

**Review by MihhailSokolov — CHANGES_REQUESTED** [review]
Looks great except some minor comments

File: analyzer/quality-analyzer/README.md
Comment (MihhailSokolov) [inline analyzer/quality-analyzer/README.md:29]: I think you still need to provide a database password in the environmental variable, don't you?

File: analyzer/metadata-plugin/src/main/java/eu/fasten/analyzer/metadataplugin/Main.java
Comment (MihhailSokolov) [inline analyzer/metadata-plugin/src/main/java/eu/fasten/analyzer/metadataplugin/Main.java:65]: I think it would be a bit more convenient to use `forge` as a key in the map of DSL Contexts instead of language. Because during runtime you choose the right context based on the forge you receive, not language. But if you prefer having a language there, I don't really have a problem with that.

File: analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/MetadataUtils.java
Comment (MihhailSokolov) [inline analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/MetadataUtils.java:68]: We already have a constants class which you can use. It's in the `core/data/Constants` and also have `mvnForge`.

File: analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/MetadataUtils.java
Comment (MihhailSokolov) [inline analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/MetadataUtils.java:193]: Package version with ID = -1 is also already taken for external callable reference. Same goes for modules and packages.
```sql
fasten_java=> select * from packages where id = -1;
 -1 | external_callables_library | mvn   |              |            | 

fasten_java=> select * from package_versions where id = -1;
 -1 |         -1 | 0.0.1   | OPAL         |            |          | 

fasten_java=> select * from modules where id=-1;
 -1 |                 -1 | global_external_callables |            | 
```

File: analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/QualityAnalyzerPlugin.java
Comment (MihhailSokolov) [inline analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/QualityAnalyzerPlugin.java:57]: Please make `utils` static and here use `QualityAnalyzer.utils = new MetadataUtils(dslContexts)`. Otherwise you will get a `NullPointerException` when running this plugin with FASTEN Server

File: analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/QualityAnalyzerPlugin.java
Comment (MihhailSokolov) [inline analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/QualityAnalyzerPlugin.java:86]: No need. The way you have it with that if-else it just find

File: analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/QualityAnalyzerPlugin.java
Comment (MihhailSokolov) [inline analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/QualityAnalyzerPlugin.java:131]: You should add a field something like `pluginError` which you return here. Then wrap everything in `consume` method in try-catch block which catches all `RuntimeException`, logs them, and sets that `pluginError` equal to the thrown exception.

File: analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/QualityAnalyzerPlugin.java
Comment (MihhailSokolov) [inline analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/QualityAnalyzerPlugin.java:89]: Also, I think that all inserts should be done within one transaction because otherwise, we might end up with partial data.
You can see how transaction management with retries works [here](https://github.com/fasten-project/fasten/blob/develop/analyzer/metadata-plugin/src/main/java/eu/fasten/analyzer/metadataplugin/MetadataDatabasePlugin.java#L140)

File: analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/data/QAConstants.java
Comment (MihhailSokolov) [inline analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/data/QAConstants.java:27]: I think that these 3 forges should be placed in that `Constants` class we have in the `core/data` because they will be used by multiple plugins. And Maven forge is already there.
Other constants can be here as they are.

**Review by MihhailSokolov — CHANGES_REQUESTED** [review]
Overall, looks good. Nice work! However, I have some comments for improvements :)

File: analyzer/quality-analyzer/README.md
Comment (zikanl) [reply analyzer/quality-analyzer/README.md:29]: Yes, that is correct, will update the documentation.

File: analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/MetadataUtils.java
Comment (zikanl) [reply analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/MetadataUtils.java:68]: Correct, I have already updated QAConstants and synched with Constants. However, I wait for the final decision about the forges, before updating it altogether. BTW, Constants do not have any reference to either C or Python forge.

File: analyzer/metadata-plugin/src/main/java/eu/fasten/analyzer/metadataplugin/Main.java
Comment (zikanl) [reply analyzer/metadata-plugin/src/main/java/eu/fasten/analyzer/metadataplugin/Main.java:65]: I didn't implement this piece of code, simply merged multiple-db-conc, before finalizing QA code.

File: analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/QualityAnalyzerPlugin.java
Comment (zikanl) [reply analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/QualityAnalyzerPlugin.java:57]: OK, done.

File: analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/QualityAnalyzerPlugin.java
Comment (zikanl) [reply analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/QualityAnalyzerPlugin.java:86]: OK, removed TODO

File: analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/data/QAConstants.java
Comment (zikanl) [reply analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/data/QAConstants.java:27]: I agree, the comment says it all (`//needs to be integrated with core.data.Constants`)

File: analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/QualityAnalyzerPlugin.java
Comment (zikanl) [reply analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/QualityAnalyzerPlugin.java:57]: OK, done

File: analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/MetadataUtils.java
Comment (zikanl) [reply analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/MetadataUtils.java]: OK, done.

File: analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/data/QAConstants.java
Comment (zikanl) [reply analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/data/QAConstants.java]: OK, done

File: analyzer/quality-analyzer/README.md
Comment (zikanl) [reply analyzer/quality-analyzer/README.md]: OK, done

File: analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/MetadataUtils.java
Comment (zikanl) [reply analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/MetadataUtils.java:189]: OK, done. Null is returned instead of -1L, and checked on it.

File: analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/QualityAnalyzerPlugin.java
Comment (zikanl) [reply analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/QualityAnalyzerPlugin.java]: OK, done

File: analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/QualityAnalyzerPlugin.java
Comment (zikanl) [reply analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/QualityAnalyzerPlugin.java]: I followed closely the implementation of the VulnerabilityConsumer - I do only updates of the metadata in the callables (using updateCallableMetadata from MetadataDao). Therefore, I have changed the function names to better illustrate this.

Comment (MihhailSokolov) [reply analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/QualityAnalyzerPlugin.java]: As long as you are certain that we won't be left with partial data in the case when your plugin unexpectedly terminates in the middle of processing, I am fine with what you have. But if there is a chance that part of the data will be written and the other part won't, then proper transaction management should be implemented.

File: analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/QualityAnalyzerPlugin.java
Comment (MihhailSokolov) [inline analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/QualityAnalyzerPlugin.java:97]: Please here add something like `this.pluginError = new RuntimeError("Could not extract forge from the message");`. Otherwise, when there is no forge, it will skip the record, and it will seem that it was processed successfully (no errors) while in reality, it wasn't successful at all and there was an error (missing forge).

Comment (zikanl) [reply analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/QualityAnalyzerPlugin.java]: OK done, with a little bit refactoring and setting pluginError to RuntimeException().

File: analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/QualityAnalyzerPlugin.java
Comment (zikanl) [reply analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/QualityAnalyzerPlugin.java]: OK, done, I refactored the suggested code. BTW, the code from MetadataDatabasePlugin should be refactored and simplified IMO.

Comment (MihhailSokolov) [reply analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/QualityAnalyzerPlugin.java]: Thanks. Yes, it is already being refactored by Stefanos (AUEB) to allow the processing of C and Python callgraphs.

File: analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/QualityAnalyzerPlugin.java
Comment (MihhailSokolov) [inline analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/QualityAnalyzerPlugin.java:103]: If you are trying to use transactions, then you are not actually starting a transaction. Here is the link that nicely explains how JOOQ transaction management works https://www.jooq.org/doc/3.12/manual/sql-execution/transaction-management/
If you don't want transactions but only retries, then you don't need that empty catch block. It is needed for proper error management in transactions (you can take a closer look at how this works in the `MetadataDatabasePlugin` class - transaction starts with lambda function, there is an inner catch block which decides whether to restart transaction or not and it always throws caught exception back from the transaction lambda to stop the transaction, then outer empty catch block catches everything but doesn't do anything because everything was already done in the inner catch block, and then transaction is restarted if needed).

Comment (zikanl) [reply analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/QualityAnalyzerPlugin.java:103]: OK, the latest version is there. I moved transactions to `MetadataUtils`, in order (hopefully) to get a more readable code, and hopefully I handle all Exceptions. I still need to throw `JsonParsingException` instead of `IllegalStateException` where applicable.

Comment (MihhailSokolov) [reply analyzer/quality-analyzer/src/main/java/eu/fasten/analyzer/qualityanalyzer/QualityAnalyzerPlugin.java:103]: Yes, the transaction management looks correct now. Thanks!

**Review by MihhailSokolov — APPROVED** [review]
Everything looks good to me, nice work. If this is completely finished, @zikanl you can merge it.

**Pull request merged**
